### ISSUE TITLE
Login loop (temporarily fixed by restarting app service)

### ISSUE THREAD

**Issue by jasonshave** [Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2#100]
### This issue is for a: (mark with an `x`)
```
- [ X ] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
> If I restart my app service I can log in just fine. Over time though when I visit the website I'm prompted for my credentials, I authenticate, then the page redirects back and forth between the 'authorize' endpoint at login.microsoftonline.com and my own signin-oidc endpoint. It does this for about 1 minute then stops.

### Any log messages given by the failure
> No log messages that I can see at all other than when the login loop times out (after 1 minute) I get "We couldn't sign you in".

### Expected/desired behavior
>

### OS and Version?
> Windows 10 dev machine. Remote server is running in Azure under an app service plan.

### Versions
> .NET Core 2.2

### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

**Comment by jmprieur** [comment]
@jasonshave : what are the repro steps? just use the sample? which sub folder?

**Comment by jasonshave** [comment]
The step to reproduce (for me at least) is to simply wait about 12 hours. The website eventually won't authenticate and loops as mentioned above. I tried setting the token type to `.AddInMemoryPerUserTokenCache` and while this seemed to work okay over the past few days, I have the login loop showing up again.

The code I'm using is from the sample [here](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/2-WebApp-graph-user/2-1-Call-MSGraph). 

I'm happy to set up a LiveShare or Teams session to show you. You can find me in the GAL.

**Comment by jmprieur** [comment]
@kalyankrishna1 @TiagoBrenck 
does one of you have time to have a look?
otherwise I'll try to get to this by EOW

**Comment by TiagoBrenck** [comment]
I will take a look at this. @jmprieur

**Comment by TiagoBrenck** [comment]
@jasonshave Could you reproduce this bug running on localhost? Or it just happens when you deploy the web app on Azure?

**Comment by ferronsw** [comment]
I'm having te same problem on a Linux Azure Web App.

**Comment by ferronsw** [comment]
With DataProtection on it still happens.

**Comment by ferronsw** [comment]
@jasonshave I can confirm this also happens when running it for 12 hours on localhost.

**Comment by TiagoBrenck** [comment]
@ferronsw I could reproduce it locally. We are still investigating it, but so far here is my analysis:

What is happening is: 
1. After X-hrs (I set locally to expire in 1 min), the token that we have cached expires. 
2. The user calls Profile Action Result, (with **MsalUiRequiredExceptionFilter**) and an access token will try to be acquired. The method `GetAccessTokenOnBehalfOfUser` (inside `TokenAcquisition.cs`) is called
3. This method builds a new `IConfidentialClientApplication` (where cache accessor is empty), receives an account identifier and tries to retrieve an `IAccount` object. It returns null because it got expired and cache accessor is empty.
4. The method calls `AcquireTokenSilent` with the null `IAccount` in the parameters. A `MsalUiRequiredException` is thrown and the filter gets in action.
5. The filter calls `BuildAuthenticationPropertiesForIncrementalConsent` that basically populates `AuthenticationProperties` with the scope, loginHint and extra claims (if needed) with the same ones that got requested previously. Then, a new `ChallengeResult` is created  with the `AuthenticationProperties` created.
6. The event `OnAuthorizationCodeReceived` gets triggered and the method `AddAccountToCacheFromAuthorizationCode` is called. 
7. A new `IConfidentialClientApplication` is built and `AcquireTokenByAuthorizationCode` is called right after. This method populates the cache accessor properly however this new `IConfidentialClientApplication` object is lost (in my opinion), because after this step, we get back to **step 2**, and all this flow happens again, and again. I would expect that on step 3, the cache accessor had the values populated by `AcquireTokenByAuthorizationCode`.


If you want to reproduce this behavior locally, download the sample https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/2-WebApp-graph-user/2-1-Call-MSGraph
Then modify the class MSALPerUserMemoryTokenCacheProvider.cs, to this one:

```csharp
        public MSALPerUserMemoryTokenCacheProvider(IMemoryCache cache, MSALMemoryTokenCacheOptions option)
        {
            this.memoryCache = cache;

            if (option != null)
            {
                this.CacheOptions = new MSALMemoryTokenCacheOptions() { AbsoluteExpiration = System.DateTime.Now.AddMinutes(1) };
            }
            else
            {
                option.AbsoluteExpiration = System.DateTime.Now.AddMinutes(1);

                this.CacheOptions = option;
            }
        }
```
Run the project, login and go to Profile page. Wait your token expiration time, then click on the Profile link again. You will then be in the loop.

**The problem**
The job done on `IConfidentialClientApplication` at step 7, doesnt affect the `IConfidentialClientApplication` created at step 2
